File: scripts/build-tesla-engram.py
#!/usr/bin/env python3
"""
Build tesla.engram — Nikola Tesla's complete published works as a neural memory file.
Chunks text semantically and ingests via Allo CLI.
"""
import os
import re
import subprocess
import sys

CORPUS_DIR = os.path.expanduser("~/clawd/allo/tesla-corpus")
OUTPUT_FILE = os.path.expanduser("~/tesla.engram")
CHUNK_SIZE = 400  # words per chunk (sweet spot for embeddings)
OVERLAP = 50  # word overlap between chunks

# Source files with metadata
SOURCES = [
    # (filename, source_label, tags)
    ("my-inventions-full.txt", "My Inventions (1919 Autobiography)", "autobiography,personal,inventions"),
    ("increasing-human-energy.txt", "The Problem of Increasing Human Energy (1900)", "essay,energy,philosophy,science"),
    ("on-light-1893.txt", "On Light and Other High Frequency Phenomena (1893)", "lecture,physics,light,high-frequency"),
    ("Nikola_Tesla.grok.txt", "Nikola Tesla — Biography", "biography,life,career"),
    ("Tesla_coil.grok.txt", "Tesla Coil — Technical Reference", "tesla-coil,invention,high-voltage"),
    ("Alternating_current.grok.txt", "Alternating Current — Technical Reference", "ac,electricity,power"),
    ("Wardenclyffe_Tower.grok.txt", "Wardenclyffe Tower — Project History", "wardenclyffe,wireless,energy-transmission"),
    ("Tesla_turbine.grok.txt", "Tesla Turbine — Technical Reference", "turbine,mechanical,invention"),
    ("Polyphase_system.grok.txt", "Polyphase System — Technical Reference", "polyphase,ac,motors,power"),
    ("Rotating_magnetic_field.grok.txt", "Rotating Magnetic Field — Technical Reference", "magnetic-field,ac,motors"),
    ("Teslas_oscillator.grok.txt", "Tesla's Oscillator — Technical Reference", "oscillator,mechanical,resonance"),
    ("Induction_motor.grok.txt", "Induction Motor — Technical Reference", "induction-motor,ac,invention"),
    ("List_of_Nikola_Tesla_patents.grok.txt", "Tesla Patents — Complete List", "patents,inventions,legal"),
    ("Nikola_Tesla_in_popular_culture.grok.txt", "Tesla in Popular Culture", "culture,legacy,media"),
]


def clean_text(text: str) -> str:
    """Remove wiki markup, HTML artifacts, and noise."""
    # Remove reference numbers like [1], [2]
    text = re.sub(r'\[\d+\]', '', text)
    # Remove image references
    text = re.sub(r'!\[.*?\]\[.*?\]', '', text)
    # Remove CSS/HTML artifacts
    text = re.sub(r'\.mw-parser-output[^{]*\{[^}]+\}', '', text)
    # Remove navigation/menu artifacts
    text = re.sub(r'(Main menu|Jump to content|Search|Appearance|Donate|Create account|Log in|Personal tools)[\s\n]*', '', text)
    # Remove repeated whitespace
    text = re.sub(r'\n{3,}', '\n\n', text)
    text = re.sub(r' {2,}', ' ', text)
    return text.strip()


def chunk_text(text: str, source_label: str, chunk_size: int = CHUNK_SIZE, overlap: int = OVERLAP):
    """Split text into overlapping chunks by paragraph boundaries."""
    paragraphs = [p.strip() for p in text.split('\n\n') if p.strip() and len(p.strip()) > 50]
    
    chunks = []
    current_words = []
    current_para_start = 0
    
    for para in paragraphs:
        words = para.split()
        current_words.extend(words)
        
        while len(current_words) >= chunk_size:
            chunk = ' '.join(current_words[:chunk_size])
            chunks.append(f"[{source_label}] {chunk}")
            current_words = current_words[chunk_size - overlap:]
    
    # Don't forget the last chunk
    if current_words and len(current_words) > 30:
        chunk = ' '.join(current_words)
        chunks.append(f"[{source_label}] {chunk}")
    
    return chunks


def main():
    all_chunks = []
    
    print("=" * 60)
    print("  Building tesla.engram — The Mind of Nikola Tesla")
    print("=" * 60)
    print()
    
    for filename, label, tags in SOURCES:
        fpath = os.path.join(CORPUS_DIR, filename)
        if not os.path.exists(fpath):
            print(f"  SKIP: {filename} (not found)")
            continue
        
        with open(fpath) as f:
            text = f.read()
        
        if len(text) < 1000:
            print(f"  SKIP: {filename} (too small: {len(text)} chars)")
            continue
        
        text = clean_text(text)
        chunks = chunk_text(text, label)
        all_chunks.extend([(c, tags) for c in chunks])
        print(f"  {filename}: {len(text):,} chars -> {len(chunks)} chunks")
    
    print(f"\n  Total: {len(all_chunks)} chunks to ingest")
    print(f"  Estimated time: ~{len(all_chunks) * 2}s ({len(all_chunks) * 2 // 60}m)\n")
    
    # Delete old file if exists
    if os.path.exists(OUTPUT_FILE):
        os.remove(OUTPUT_FILE)
        print(f"  Removed old {OUTPUT_FILE}")
    
    # Ingest via allo remember
    success = 0
    errors = 0
    for i, (chunk, tags) in enumerate(all_chunks):
        if i % 25 == 0:
            pct = (i / len(all_chunks)) * 100
            print(f"  [{i}/{len(all_chunks)}] ({pct:.0f}%) — {success} ingested, {errors} errors")
        
        # Escape quotes for shell
        safe_chunk = chunk.replace("'", "'\\''")
        safe_tags = tags.replace("'", "'\\''")
        
        cmd = f"allo remember '{safe_chunk}' --tags '{safe_tags}' --file '{OUTPUT_FILE}'"
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=60)
        
        if result.returncode == 0:
            success += 1
        else:
            errors += 1
            if errors <= 3:
                print(f"    ERROR at chunk {i}: {result.stderr[:100]}")
    
    print(f"\n{'=' * 60}")
    print(f"  Done! {success}/{len(all_chunks)} chunks ingested")
    print(f"  File: {OUTPUT_FILE}")
    print(f"{'=' * 60}")
    
    # Show stats
    subprocess.run(f"allo stats --file '{OUTPUT_FILE}'", shell=True)


if __name__ == "__main__":
    main()

File: scripts/build-tesla-engram.js
#!/usr/bin/env node
/**
 * Build tesla.engram — Nikola Tesla's complete published works.
 * Chunks text semantically, embeds, and saves to a single .engram file.
 */
const { Allo } = require('../dist/allo');
const fs = require('fs');
const path = require('path');

const CORPUS_DIR = path.join(__dirname, '..', 'tesla-corpus');
const OUTPUT_FILE = path.join(require('os').homedir(), 'tesla.engram');
const CHUNK_SIZE = 400; // words
const OVERLAP = 50;

const SOURCES = [
    // Primary sources — Tesla's own words
    ["my-inventions-full.txt", "My Inventions (1919 Autobiography)", "autobiography,personal,first-person"],
    ["increasing-human-energy.txt", "The Problem of Increasing Human Energy (1900)", "essay,energy,philosophy,first-person"],
    ["on-light-1893.txt", "On Light and High Frequency Phenomena (1893)", "lecture,physics,light,first-person"],
    ["The_True_Wireless.wiki.txt", "The True Wireless (1919)", "essay,wireless,radio,first-person"],
    // Comprehensive biography
    ["Nikola_Tesla.grok.txt", "Tesla Biography", "biography,life,career"],
    // Technical inventions
    ["Tesla_coil.grok.txt", "Tesla Coil", "tesla-coil,invention,high-voltage"],
    ["Alternating_current.grok.txt", "Alternating Current", "ac,electricity,power"],
    ["Polyphase_system.grok.txt", "Polyphase System", "polyphase,ac,motors"],
    ["Rotating_magnetic_field.grok.txt", "Rotating Magnetic Field", "magnetic-field,motors"],
    ["Induction_motor.grok.txt", "Induction Motor", "induction-motor,ac"],
    ["Tesla_turbine.grok.txt", "Tesla Turbine", "turbine,mechanical"],
    ["Teslas_oscillator.grok.txt", "Tesla's Oscillator", "oscillator,resonance"],
    // Wireless & energy transmission
    ["Wardenclyffe_Tower.grok.txt", "Wardenclyffe Tower", "wardenclyffe,wireless,energy"],
    ["World_Wireless_System.grok.txt", "World Wireless System", "wireless,global,communication"],
    ["Wireless_power_transfer.grok.txt", "Wireless Power Transfer", "wireless,energy,transmission"],
    ["Tesla_Experimental_Station.grok.txt", "Colorado Springs Laboratory", "colorado-springs,experiments,lab"],
    // Weapons & defense
    ["Teleforce.grok.txt", "Teleforce (Death Ray)", "teleforce,weapon,defense"],
    ["Death_ray.grok.txt", "Death Ray History", "death-ray,weapon,military"],
    // Legacy & history
    ["History_of_radio.grok.txt", "History of Radio", "radio,wireless,marconi"],
    ["Tesla_Science_Center_at_Wardenclyffe.grok.txt", "Tesla Science Center", "museum,legacy,wardenclyffe"],
    ["Nikola_Tesla_Museum.grok.txt", "Nikola Tesla Museum", "museum,legacy,belgrade"],
    ["Nikola_Tesla_in_popular_culture.grok.txt", "Tesla in Popular Culture", "culture,legacy,media"],
    // Patents
    ["List_of_Nikola_Tesla_patents.grok.txt", "Tesla Patents", "patents,inventions,legal"],
];

function cleanText(text) {
    text = text.replace(/\[\d+\]/g, '');
    text = text.replace(/!\[.*?\]\[.*?\]/g, '');
    text = text.replace(/\.mw-parser-output[^{]*\{[^}]+\}/g, '');
    text = text.replace(/(Main menu|Jump to content|Search|Appearance|Donate|Create account|Log in|Personal tools)[\s\n]*/g, '');
    text = text.replace(/\n{3,}/g, '\n\n');
    text = text.replace(/ {2,}/g, ' ');
    return text.trim();
}

function chunkText(text, sourceLabel) {
    const paragraphs = text.split('\n\n').filter(p => p.trim().length > 50);
    const chunks = [];
    let currentWords = [];

    for (const para of paragraphs) {
        currentWords.push(...para.split(/\s+/));

        while (currentWords.length >= CHUNK_SIZE) {
            const chunk = currentWords.slice(0, CHUNK_SIZE).join(' ');
            chunks.push(`[${sourceLabel}] ${chunk}`);
            currentWords = currentWords.slice(CHUNK_SIZE - OVERLAP);
        }
    }

    if (currentWords.length > 30) {
        chunks.push(`[${sourceLabel}] ${currentWords.join(' ')}`);
    }

    return chunks;
}

async function main() {
    console.log('='.repeat(60));
    console.log('  Building tesla.engram — The Mind of Nikola Tesla');
    console.log('='.repeat(60));
    console.log();

    // Collect all chunks
    const allChunks = [];
    for (const [filename, label, tags] of SOURCES) {
        const fpath = path.join(CORPUS_DIR, filename);
        if (!fs.existsSync(fpath)) {
            console.log(`  SKIP: ${filename} (not found)`);
            continue;
        }
        const text = cleanText(fs.readFileSync(fpath, 'utf-8'));
        if (text.length < 1000) {
            console.log(`  SKIP: ${filename} (too small)`);
            continue;
        }
        const chunks = chunkText(text, label);
        allChunks.push(...chunks.map(c => ({ text: c, tags: tags.split(',') })));
        console.log(`  ${filename}: ${text.length.toLocaleString()} chars -> ${chunks.length} chunks`);
    }

    console.log(`\n  Total: ${allChunks.length} chunks to ingest`);

    // Remove old file
    if (fs.existsSync(OUTPUT_FILE)) {
        fs.unlinkSync(OUTPUT_FILE);
        console.log(`  Removed old ${OUTPUT_FILE}`);
    }

    // Initialize Allo once (keeps embedding model warm)
    console.log('\n  Initializing Allo...');
    const allo = new Allo({ memoryFile: OUTPUT_FILE });
    await allo.initialize();
    console.log('  Allo ready!\n');

    // Ingest all chunks
    let success = 0;
    let errors = 0;
    const startTime = Date.now();

    for (let i = 0; i < allChunks.length; i++) {
        const { text, tags } = allChunks[i];

        if (i % 25 === 0) {
            const pct = ((i / allChunks.length) * 100).toFixed(0);
            const elapsed = ((Date.now() - startTime) / 1000).toFixed(0);
            console.log(`  [${i}/${allChunks.length}] (${pct}%) — ${success} ok, ${errors} err — ${elapsed}s`);
        }

        try {
            await allo.addText(text, undefined, tags);
            success++;
        } catch (e) {
            errors++;
            if (errors <= 3) console.log(`    ERROR at ${i}: ${e.message}`);
        }
    }

    // Save
    console.log('\n  Saving tesla.engram...');
    const { nodeCount, fileSizeMB } = await allo.save();

    const totalTime = ((Date.now() - startTime) / 1000).toFixed(0);
    console.log(`\n${'='.repeat(60)}`);
    console.log(`  Done! ${success}/${allChunks.length} chunks ingested in ${totalTime}s`);
    console.log(`  File: ${OUTPUT_FILE}`);
    console.log(`  Size: ${fileSizeMB} MB | Memories: ${nodeCount}`);
    console.log(`${'='.repeat(60)}`);
}

main().catch(e => { console.error(e); process.exit(1); });
